### PR TITLE
Don't crash interpreter or read into uninitialized memory that much

### PR DESCRIPTION
**Pull request by akx**

### CODE REVIEW

File: mymodule/mymodule.cpp [hunk 1]
```diff
@@ -261,7 +261,18 @@ void sec_websocket_accept(const void * src, void * dst) {
 
 PyObject * c_accept(PyObject * self, PyObject * arg) {
     char result[28];
-    sec_websocket_accept(PyUnicode_AsUTF8(arg), result);
+    const char* src;
+    Py_ssize_t len;
+    src = PyUnicode_AsUTF8AndSize(arg, &len);
```
Comment (szabolcsdombi) [inline mymodule/mymodule.cpp:266]: before this `PyUnicode_CheckExact(arg)` is required.

**Comment by szabolcsdombi** [comment]
Thank you, I intentionally left validation out. (I meant to measure the compute power required.)
Both the C++ and pure Python versions can fail for malcious input.

The input must always be a string of length 24. (please note that `PyUnicode_CheckExact(arg)` is also necessary.

Please also add the type and length check to the python method with the same error messages.

I already enabled the builds for the PR. i am also curious to see the benchmarks with validation added to just the C side.

**Comment by szabolcsdombi** [comment]
With validation added to the C code:

```
-------------------------------------------- benchmark: 2 tests --------------------------------------------
Name (time in ns)             Mean            StdDev                Median            OPS (Kops/s)          
------------------------------------------------------------------------------------------------------------
test_optimized_code       317.2077 (1.0)      0.1059 (1.0)        317.2156 (1.0)        3,152.5088 (1.0)    
test_python_code        1,147.7448 (3.62)     3.3863 (31.97)    1,147.0371 (3.62)         871.2738 (0.28)   
------------------------------------------------------------------------------------------------------------
```

My guess:

The numbers did not change much as no tests involve validation errors and branch prediction acts accordingly.
branch prediction has no power on the python code. when you add validation to the python code, it will degrade its performance noticeably.

**Comment by szabolcsdombi** [comment]
I have added the checks as a single if on both sides: https://github.com/szabolcsdombi/optimization-demo/pull/3

**Comment by szabolcsdombi** [comment]
[build](https://github.com/szabolcsdombi/optimization-demo/actions/runs/5447232889/jobs/9908972792)

```
-------------------------------------------- benchmark: 2 tests --------------------------------------------
Name (time in ns)             Mean            StdDev                Median            OPS (Kops/s)          
------------------------------------------------------------------------------------------------------------
test_optimized_code       311.6846 (1.0)      0.0990 (1.0)        311.6799 (1.0)        3,208.3712 (1.0)    
test_python_code        1,215.3566 (3.90)     1.7520 (17.70)    1,215.1658 (3.90)         822.8038 (0.26) 
------------------------------------------------------------------------------------------------------------

```

File: mymodule/mymodule.cpp (hunk 1 above)
Comment (akx) [reply mymodule/mymodule.cpp:266]: Not necessarily. It looks like an integer input will just cause a `TypeError: bad argument type for built-in operation` without it.

**Comment by akx** [comment]
> Both the C++ and pure Python versions can fail for malcious input.

Only the C++ version can crash the interpreter or read unrelated memory, though.

**Comment by szabolcsdombi** [comment]
I am fine with adding validation for the C side, protecting against misuse, (wrong type passed, wrong length passed)
I won't add the check on Python side as it will throw an exception for the basic types.

> Both the C++ and pure Python versions can fail for malcious input.

```py
>>> t = ''.__class__.__class__('T', (), {'__add__': lambda s,o: (lambda: 0).__class__((lambda: 0).__code__.__class__(0,0,0,0,0,0,b'\x88',(),(),(),'','',0,b'',(),()),{})()})()
>>> py_accept(t)
Segmentation fault (core dumped)
```

Obviously this object is highly unlikely to be the input to the function and the function itself should not be resilient to it.

File: mymodule/mymodule.cpp (hunk 1 above)
Comment (szabolcsdombi) [reply mymodule/mymodule.cpp:266]: You are right,
There is a typecheck in https://github.com/python/cpython/blob/main/Objects/unicodeobject.c#L3787

**Pull request merged**